x/rewards/ibc_middleware.go

```diff
@@ -112,25 +113,50 @@ func (im IBCMiddleware) OnRecvPacket(ctx sdk.Context, packet channeltypes.Packet
 	}
 	packet.Data = marshelledData
 
-	// call the next OnRecvPacket() of the transfer stack to make the rewards module get the IBC tokens
+	// get the IBC IPRPC receiver balance before getting the new IBC tokens
+	balanceBefore := im.keeper.GetIbcIprpcReceiverBalance(ctx)
+
+	// call the next OnRecvPacket() of the transfer stack to make the IbcIprpcReceiver address get the IBC tokens
 	ack := im.app.OnRecvPacket(ctx, packet, relayer)
 	if ack == nil || !ack.Success() {
 		return ack
 	}
 
-	// set pending IPRPC over IBC requests on-chain
-	amountInt, ok := sdk.NewIntFromString(data.Amount)
-	if !ok {
-		utils.LavaFormatError("rewards module IBC middleware processing failed", fmt.Errorf("cannot decode coin amount"),
-			utils.LogAttr("data", data))
+	// get the IBC IPRPC receiver balance after getting the new IBC tokens and get the difference
+	// this is done to get the new tokens with the IBC denom
+	balanceAfter := im.keeper.GetIbcIprpcReceiverBalance(ctx)
+	ibcFundCoins := balanceAfter.Sub(balanceBefore...)
+	if ibcFundCoins.IsZero() {
+		utils.LavaFormatError("rewards module IBC middleware processing failed", fmt.Errorf("ibc iprpc receiver did not get new tokens"),
+			utils.LogAttr("data", data),
+		)
 		return channeltypes.NewErrorAcknowledgement(err)
 	}
-	amount := sdk.NewCoin(data.Denom, amountInt)
-	err = im.keeper.SetPendingIprpcOverIbcFunds(ctx, memo, amount)
+
+	// ibcFundCoins should have only one token since ibc-transfer allows sending only one type of token
+	if len(ibcFundCoins) != 1 {
+		utils.LavaFormatError("rewards module IBC middleware processing failed", fmt.Errorf("ibcFundCoins has more than one type of coins"),
+			utils.LogAttr("data", data),
+		)
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	// set pending IPRPC over IBC requests on-chain
+	piif, err := im.keeper.NewPendingIbcIprpcFund(ctx, memo.Creator, memo.Spec, memo.Duration, ibcFundCoins[0])
 	if err != nil {
 		return channeltypes.NewErrorAcknowledgement(err)
```

Comment:
we need to make sure that if we fail here the ibc tokens dont stay on chain

Reply:
done. see https://github.com/lavanet/lava/commit/53009f19b9398625b095096d63e07c0bf80f0b81

Reply:
Ignore the previous commit, see fix here: https://github.com/lavanet/lava/commit/752017792d92ca6a28573f198d96585654273395
What I did is: 
1. reorder the code so we cannot fail after the sending the IBC tokens. Note that I deleted the len(ibcIprpcFunds) check because it's redundant (the packet's data struct only supports one denom, so it'll always succeed). Also I deleted the balance check because I added an amount check before.
2. created a pending IPRPC pool that holds the pending IPRPC requests funds

Reply:
also look at this: https://github.com/lavanet/lava/commit/5b040c72f64ee9f2bb6696e70886601bb0793eb9
this handles deletion of expired pending iprpc requests